src/main/java/com/minecolonies/coremod/inventory/CraftingGUIBuilding.java

```diff
@@ -0,0 +1,234 @@
+
+    /**
+     * The world object.
+     */
+    private final World worldObj;
+
+    public CraftingGUIBuilding(final InventoryPlayer playerInventory, final World worldIn, final BlockPos posIn)
+    {
+        super();
+        this.worldObj = worldIn;
+
+        this.addSlotToContainer(new SlotCrafting(playerInventory.player, this.craftMatrix, this.craftResult, 0, X_CRAFT_RESULT, Y_CRAFT_RESULT));
+
+        for (int i = 0; i < 2; ++i)
+        {
+            for (int j = 0; j < 2; ++j)
+            {
+                this.addSlotToContainer(new Slot(this.craftMatrix, j + i * 2, X_OFFSET_CRAFTING + j * INVENTORY_OFFSET_EACH, Y_OFFSET_CRAFTING + i * INVENTORY_OFFSET_EACH));
+            }
+        }
+
+        int i;
+        for (i = 0; i < PLAYER_INVENTORY_ROWS; i++)
+        {
+            for (int j = 0; j < PLAYER_INVENTORY_COLUMNS; j++)
+            {
+                addSlotToContainer(new Slot(
+                        playerInventory,
+                        j + i * PLAYER_INVENTORY_COLUMNS + PLAYER_INVENTORY_COLUMNS,
+                        PLAYER_INVENTORY_INITIAL_X_OFFSET + j * INVENTORY_OFFSET_EACH,
+                        PLAYER_INVENTORY_INITIAL_Y_OFFSET + i * INVENTORY_OFFSET_EACH
+                ));
+            }
+        }
+
+        for (i = 0; i < PLAYER_INVENTORY_COLUMNS; i++)
+        {
+            addSlotToContainer(new Slot(
+                    playerInventory, i,
+                    PLAYER_INVENTORY_INITIAL_X_OFFSET + i * INVENTORY_OFFSET_EACH,
+                    PLAYER_INVENTORY_HOTBAR_OFFSET
+            ));
+        }
+
+        this.onCraftMatrixChanged(this.craftMatrix);
+    }
+
+    @Override
+    protected final Slot addSlotToContainer(final Slot slotToAdd)
+    {
+        return super.addSlotToContainer(slotToAdd);
+    }
+
+    /**
+     * Callback for when the crafting matrix is changed.
+     */
+    @Override
+    public void onCraftMatrixChanged(IInventory inventoryIn)
+    {
+        super.onCraftMatrixChanged(inventoryIn);
+        //todo can use this to get the recipe on button click
+        this.craftResult.setInventorySlotContents(0, CraftingManager.getInstance().findMatchingRecipe(this.craftMatrix, this.worldObj));
+    }
+
+    /**
+     * Called when the container is closed.
+     */
+    @Override
+    public void onContainerClosed(EntityPlayer playerIn)
+    {
+        super.onContainerClosed(playerIn);
+
+        if (!this.worldObj.isRemote)
+        {
+            for (int i = 0; i < 9; ++i)
+            {
+                ItemStack itemstack = this.craftMatrix.removeStackFromSlot(i);
+
+                if (itemstack != null)
+                {
+                    playerIn.dropItem(itemstack, false);
+                }
+            }
+        }
+    }
+
+    @Override
+    public boolean canInteractWith(EntityPlayer playerIn)
+    {
+        return true;
+    }
+
+
+    @Nullable
+    @Override
+    public ItemStack transferStackInSlot(EntityPlayer playerIn, int index)
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 23 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)